File: app/src/main/java/rrzaniolo/testandroidsantander/network/models/ui/base/CellView.java
package rrzaniolo.testandroidsantander.network.models.ui.base;

/*
 * Created by Rodrigo Rodrigues Zaniolo on 5/8/2018.
 * All rights reserved.
 */

import android.support.annotation.LayoutRes;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import rrzaniolo.testandroidsantander.network.models.data.Cell;

/***
 * View representation of the Cell model.
 * */
public abstract class CellView {

    //region --- Variables
    private Cell cell;
    public Cell getCell() {
        return cell;
    }
    public void setCell(Cell cell) {
        this.cell = cell;
    }

    private View view;
    public View getView() {
        return view;
    }
    public void setView(View view) {
        this.view = view;
    }

    private ViewGroup viewGroup;
    public ViewGroup getViewGroup() {
        return viewGroup;
    }
    public void setViewGroup(ViewGroup viewGroup) {
        this.viewGroup = viewGroup;
    }

    private OnShowView onShowView;
    public OnShowView getOnShowView() {
        return onShowView;
    }
    public void setOnShowView(OnShowView onShowView) {
        if(getCell() != null && getCell().getShow() != -1)
            this.onShowView = onShowView;
    }
    //endregion

    //region --- Constructor
    public CellView(Cell cell, ViewGroup viewGroup) {
        setCell(cell);
        setViewGroup(viewGroup);
    }
    //endregion

    //region --- Protected Methods
    protected void inflateLayout(@LayoutRes int layoutId){
        setView(LayoutInflater
                .from(getViewGroup().getContext())
                .inflate(layoutId, getViewGroup(), false)
        );

        setMargins();

        if(getCell().getHidden())
            getView().setVisibility(View.GONE);
    }

    //region --- Private Methods
    private void setMargins(){
        LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) getView().getLayoutParams();
        params.setMargins(0, getCell().getTopSpacing().intValue(), 0,0 );

        getView().setLayoutParams(params);
    }
    //endregion

    //region --- Methods
    protected abstract void inflateView();
    //endregion
}

File: app/src/main/java/com/vctapps/santanderchallenge/form/presentation/domain/cell/base/CellView.kt
package com.vctapps.santanderchallenge.form.presentation.domain.cell.base

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import com.vctapps.santanderchallenge.core.domain.InvalidData
import com.vctapps.santanderchallenge.form.domain.cell.Cell
import com.vctapps.santanderchallenge.form.presentation.view.OnShowViewRequest

abstract class CellView(val cell: Cell, private val rootView: ViewGroup) {

    protected lateinit var view: View

    protected var onShowViewRequest: OnShowViewRequest? = null

    abstract fun inflateView()

    fun getCellView() = view

    protected fun inflateLayout(layoutId: Int){
        this.view = LayoutInflater
                .from(rootView.context)
                .inflate(layoutId, rootView, false)

        setMargins()

        if(cell.hidden){
            this.view.visibility = View.GONE
        }
    }

    private fun setMargins(){
        val params = this.view.layoutParams as LinearLayout.LayoutParams

        params.setMargins(0, cell.topSpacing.toInt(), 0, 0)
    }

    fun setShowViewRequest(listener: OnShowViewRequest){
        if(cell.show != InvalidData.INVALID_INT){
            onShowViewRequest = listener
        }
    }

}